### PR TITLE
[Wasm GC] Casts of a non-nullable bottom type to non-null fail

### PR DESCRIPTION
**Pull request by kripken**
Casting `(ref nofunc)` to `(ref func)` seems like it can succeed based on the rule
of "if it's a subtype, it can cast ok." But the fuzzer found a corner case where that
leads to a validation error (see testcase). Instead, it's more precise anyhow to
say that such casts fail, since they will indeed fail at runtime before getting
there.

### CODE REVIEW

**Review by tlively — COMMENTED** [review]
I would have thought that considering these impossible casts to succeed would be simpler and more precise than considering them to fail.

**Comment by kripken** [comment]
Can you explain why you think that? Specifically, is there something wrong with the logic in this comment?

```cpp
    // Non-null references to bottom types do not exist, so there's no value
    // that could make the cast succeed.
```

Note that that comment was just moved around in the PR - it existed before, and so we returned Failure before. It's just that other logic before it prevented us from even reaching it.

**Comment by kripken** [comment]
I see. In that case, though, are you saying that pre-existing comment I quoted was wrong?

It does seem like as this is an impossible situation, perhaps we can handle it either way, so long as callers are aware of it. That means adding logic to OptimizeInstructions in this case, since it looks at intermediate casts, so it will need to check if any one we say is succeeding is actually possible or not (we shouldn't optimize if not).

**Comment by tlively** [comment]
No, the comment isn't wrong, it's just also true that there is no value that could make the cast fail.

I agree it would be good to avoid having to do tricky reasoning in callers, though. Can you explain the validation failure that the current code can produce in more detail?

**Comment by kripken** [comment]
Sure, the testcase has this:
```wat
    (ref.cast func
      (local.tee $0
        (loop (result (ref nofunc))
          (unreachable)
        )
      )
    )
```

We start by looking at the cast of the tee, and find nothing to do, since it's a cast of a nullable func to a non-nullable function, which might succeed or fail. Then we look further inside at the loop, and now we are casting a non-nullable nofunc to a non-nullable func, which is a subtype. Since it looks like it must always succeed, we want to replace the cast with the value, but that causes a problem here, as the only possible value is a null, but we can't replace the cast with a null.

Overall I think we can fix this in the pass by looking at the value and handling null more carefully. But given that we are talking about a cast that is never reached, we could in principle tag them as either succeeding or failing, and failing leads to more optimizations in general (a trap, that propagates out), so that seems better to me. (As a side note, this PR also makes the function more efficient, by not doing first a subtype check and then more specific checks later that overlap with it.)

**Comment by tlively** [comment]
I think the best solutions here would be to either fix OptimizeInstructions to avoid trying to materialize a value where no possible value exists, or to add a new `Unreachable` variant to `EvaluationResult` and handle that.

I expect that the solution of just changing our result from "Success" to "Failure" will lead to very similar future problems for examples that use `flipEvaluationResult`.

**Comment by kripken** [comment]
I like the idea to add an `Unreachable` option here. But I'm not sure when to use it. The code here is this:
```cpp
  if (castType.isNonNullable() && refHeapType.isBottom()) {
    // Non-null references to bottom types do not exist, so there's no value
    // that could make the cast succeed. Note that we check this before the
    // next check because bottom types are subtypes, that is, (ref nofunc)
    // is a subtype of (ref func), so it seems like the cast would succeed,
    // but we want to return failure here because that is more precise.
    return Failure;
  }
```
That condition does not actually indicate the cast is unreachable. It is checking for things like this:
```wat
(ref.cast func
  (ref.null nofunc)
)
```
A non-null cast and a bottom input. (The input might also be uninhabitable, but that's separate from what the condition is checking.)

I think returning Failure after that condition is the right thing. Perhaps separately we could also add an `Unreachable` for uninhabitable situations, but that would require more code, and I'm not sure of the benefit.

**Comment by tlively** [comment]
Unless I'm misreading the code, we already return `Failure` in that case, though. I was thinking that we would return `Unreachable` whenever and only whenever the input type is uninhabitable.

**Comment by kripken** [comment]
After sleeping on this, here is how I see things: IIUC the problem is that a non-nullable cast of a non-nullable bottom type is "ambiguous" because two valid rules apply to it:

* A non-nullable cast of a bottom type will fail.
* A cast of a subtype will succeed.

The contradiction is resolved because the input is uninhabitable so the cast is never reached. But, that also means that it doesn't matter what we return for it.

We could handle uninhabitable types in a special way, but that would be extra code that doesn't prevent an actual problem, given the above, unless I'm missing something. If I'm not, then I think a large comment explaining the issue is enough, and otherwise we can do whatever makes the code simpler and more efficient, which I think is what this PR does (as I've mentioned, it now avoids redundant subtyping work).

What do you think @tlively ?

**Comment by tlively** [comment]
As long as callers assume that `Success` means there is a concrete value that can be used to replace the cast, I don't think working around the issue by returning `Failure` in the problematic cases is an appropriate fix.

For example, it seems plausible that the `Failure` would be flipped back to `Success` and land us with the same problem we started with for code like this, if not now, then possibly in the future:

```wat
    (ref.cast func
      (local.tee $0
        (br_on_cast_fail $l null func
          (loop (result (ref nofunc))
            (unreachable)
          )
        )
      )
    )
````

I still think we need to either fix the callers to stop making bad assumptions or we need to treat these cases specially in the result of `evaluateCastCheck` (to help callers stop making bad assumptions). After we fix the logical problem, we can separately consider making `evaluateCastCheck` more efficient.

**Comment by kripken** [comment]
I think I see what you mean better now, but in practice the code won't flip it - it will look through the `br_on_cast` at the fallthrough value.

It is true the `success` path needs to be careful to not emit a null constant. But that situation cannot be reached after this PR (we declare such things as `failure`), so I think an assertion would be enough there. I'll add that.

**Comment by kripken** [comment]
@tlively updated after our chat with a forward-looking `Unreachable` case, that should avoid the potential issue you mentioned.

File: src/ir/gc-type-utils.h [hunk 1]
```diff
@@ -59,13 +69,16 @@ inline EvaluationResult flipEvaluationResult(EvaluationResult result) {
 // what we know about the result.
 inline EvaluationResult evaluateCastCheck(Type refType, Type castType) {
   if (!refType.isRef() || !castType.isRef()) {
-    // Unreachable etc. are meaningless situations in which we can inform the
-    // caller about nothing useful.
+    if (refType == Type::unreachable) {
+      return Unreachable;
+    }
+    // If the cast type is unreachable, we can't tell - perhaps this is a br
+    // instruction of some kind, that has unreachable type normally.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/gc-type-utils.h:76]: I think you're right, but we might change the encoding of these instructions in theory. As this isn't an important optimization - DCE will handle this code - I figured it's safer to be careful here.

File: src/ir/gc-type-utils.h [hunk 2]
```diff
@@ -88,17 +101,37 @@ inline EvaluationResult evaluateCastCheck(Type refType, Type castType) {
     return SuccessOnlyIfNull;
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/gc-type-utils.h]: Good point. I think callers handled that, but it's best to be precise here. I reordered it to fix that now.

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -2158,6 +2176,10 @@ struct OptimizeInstructions
         replaceCurrent(builder.makeBlock(
           {builder.makeDrop(curr->ref), builder.makeConst(int32_t(1))}));
         break;
+      case GCTypeUtils::Unreachable:
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:2171]: Hmm, good point, that would propagate out which can help. Updated. (This required removing some obsolete code that was in the way, see last commits.)

**Review by tlively — APPROVED** [review]
LGTM!

**Pull request merged**